// specky-sdd/dist/constants.js
/**
 * Specky MCP Server — Constants
 * All shared constants, enums, and configuration values.
 */
/** Specky server version — read from package.json at build time */
import { createRequire } from "node:module";
const _require = createRequire(import.meta.url);
const _pkg = _require("../package.json");
export const VERSION = _pkg.version;
/** Server name for MCP handshake */
export const SERVER_NAME = "specky";
/** Total tool count for ecosystem check */
export const TOTAL_TOOLS = 57;
/**
 * MCP Ecosystem — Recommended external MCP servers.
 * Each entry describes a server that enhances Specky's capabilities.
 * Specky works standalone, but these servers unlock integrations.
 */
export const MCP_ECOSYSTEM = [
    {
        id: "markitdown",
        name: "Microsoft MarkItDown",
        purpose: "Enhanced document conversion — PDF, DOCX, PPTX, XLSX to Markdown with superior formatting preservation. Microsoft-backed, handles complex layouts, tables, and embedded images.",
        install_command: 'Add to VS Code MCP settings: { "markitdown": { "type": "stdio", "command": "uvx", "args": ["markitdown-mcp"] } }',
        install_note: "Requires Python/uvx. Install uvx with: curl -LsSf https://astral.sh/uv/install.sh | sh",
        required: false,
        status: "recommended",
        enhances: ["sdd_import_document", "sdd_batch_import", "sdd_auto_pipeline"],
    },
    {
        id: "github",
        name: "GitHub MCP Server",
        purpose: "Create GitHub Issues, Pull Requests, and Codespaces directly from Specky outputs. Required for sdd_export_work_items(github), sdd_create_pr, and sdd_setup_codespaces.",
        install_command: 'Add to VS Code MCP settings: { "github": { "type": "http", "url": "https://api.githubcopilot.com/mcp/" } }',
        install_note: "Available in VS Code Copilot MCP Gallery. No additional setup needed if using Copilot.",
        required: false,
        status: "recommended",
        enhances: ["sdd_export_work_items", "sdd_create_pr", "sdd_create_branch", "sdd_setup_codespaces"],
    },
    {
        id: "azure-devops",
        name: "Azure DevOps MCP Server",
        purpose: "Create Work Items in Azure Boards from Specky task exports. Required for sdd_export_work_items(azure_boards).",
        install_command: 'Add to VS Code MCP settings: { "microsoft/azure-devops-mcp": { "type": "stdio", "command": "npx", "args": ["-y", "@azure-devops/mcp@latest", "<org>", "-d", "work-items"] } }',
        install_note: "Replace <org> with your Azure DevOps organization name. Requires Azure DevOps PAT or Azure AD auth.",
        required: false,
        status: "recommended",
        enhances: ["sdd_export_work_items"],
    },
    {
        id: "jira",
        name: "Jira MCP Server (Atlassian)",
        purpose: "Create Jira Issues from Specky task exports. Required for sdd_export_work_items(jira).",
        install_command: "Search for 'Jira MCP Server' in the VS Code MCP Gallery, or use: npx @anthropic/jira-mcp-server",
        install_note: "Requires Jira API token. Get from: https://id.atlassian.com/manage-profile/security/api-tokens",
        required: false,
        status: "recommended",
        enhances: ["sdd_export_work_items"],
    },
    {
        id: "terraform",
        name: "HashiCorp Terraform MCP Server",
        purpose: "Validate and plan Terraform configurations generated by Specky. Required for sdd_validate_iac(terraform).",
        install_command: 'Add to VS Code MCP settings: { "terraform": { "type": "stdio", "command": "docker", "args": ["run", "-i", "--rm", "hashicorp/terraform-mcp-server:latest"] } }',
        install_note: "Requires Docker. For HCP Terraform, set TFE_ADDRESS and TFE_TOKEN environment variables.",
        required: false,
        status: "recommended",
        enhances: ["sdd_validate_iac", "sdd_generate_iac"],
    },
    {
        id: "azure",
        name: "Azure MCP Server",
        purpose: "Validate Azure infrastructure templates (ARM/Bicep), manage resources, and deploy. Enhances sdd_validate_iac for Azure cloud.",
        install_command: 'Add to VS Code MCP settings: { "azure": { "type": "stdio", "command": "npx", "args": ["-y", "@azure/mcp@latest", "server", "start"] } }',
        install_note: "Requires Azure CLI authentication. Run: az login",
        required: false,
        status: "recommended",
        enhances: ["sdd_validate_iac", "sdd_generate_iac", "sdd_setup_local_env"],
    },
    {
        id: "figma",
        name: "Figma MCP Server",
        purpose: "Extract design context from Figma files and generate FigJam diagrams. Required for sdd_figma_to_spec and sdd_figma_diagram.",
        install_command: 'Add to VS Code MCP settings: { "figma": { "type": "http", "url": "http://127.0.0.1:3845/sse" } } OR use Figma MCP Gallery entry.',
        install_note: "Requires Figma desktop app running with MCP plugin enabled, or use the cloud-hosted Figma MCP.",
        required: false,
        status: "recommended",
        enhances: ["sdd_figma_to_spec", "sdd_figma_diagram"],
    },
    {
        id: "docker",
        name: "Docker MCP (via Docker Desktop)",
        purpose: "Create and manage local development environments with Docker Compose. Enhances sdd_setup_local_env.",
        install_command: "Install Docker Desktop from: https://www.docker.com/products/docker-desktop/",
        install_note: "Docker Desktop includes MCP support. No additional MCP server installation needed.",
        required: false,
        status: "recommended",
        enhances: ["sdd_setup_local_env", "sdd_generate_dockerfile"],
    },
    {
        id: "playwright",
        name: "Playwright MCP Server",
        purpose: "Run browser-based E2E tests against implemented features. Enhances the Verify phase.",
        install_command: 'Add to VS Code MCP settings: { "playwright": { "type": "stdio", "command": "npx", "args": ["@playwright/mcp@latest"] } }',
        install_note: "Requires Playwright browsers. Run: npx playwright install",
        required: false,
        status: "recommended",
        enhances: ["sdd_verify_tasks"],
    },
    {
        id: "miro",
        name: "Miro MCP Server",
        purpose: "Create collaborative diagrams and boards from Specky outputs. Alternative to FigJam.",
        install_command: 'Add to VS Code MCP settings: { "miro": { "type": "http", "url": "https://mcp.miro.com/" } }',
        install_note: "Requires Miro account. Available in VS Code MCP Gallery.",
        required: false,
        status: "recommended",
        enhances: ["sdd_generate_diagram", "sdd_generate_all_diagrams"],
    },
];
/** Maximum response size in characters before truncation */
export const CHARACTER_LIMIT = 25_000;
/** Default directory for SDD specifications */
export const DEFAULT_SPEC_DIR = ".specs";
/** State file name */
export const STATE_FILE = ".sdd-state.json";
/** Default HTTP port for Streamable HTTP transport */
export const DEFAULT_HTTP_PORT = 3200;
/** Default scan depth for codebase scanner */
export const DEFAULT_SCAN_DEPTH = 3;
/** Maximum scan depth allowed */
export const MAX_SCAN_DEPTH = 5;
/** Default exclude patterns for codebase scanning */
export const DEFAULT_EXCLUDE_PATTERNS = [
    "node_modules",
    ".git",
    "dist",
    ".next",
    "__pycache__",
    ".venv",
    "target",
    "build",
    "coverage",
];
/**
 * SDD Pipeline phases in order.
 * Each phase requires its predecessor to be completed.
 */
export var Phase;
(function (Phase) {
    Phase["Init"] = "init";
    Phase["Discover"] = "discover";
    Phase["Specify"] = "specify";
    Phase["Clarify"] = "clarify";
    Phase["Design"] = "design";
    Phase["Tasks"] = "tasks";
    Phase["Analyze"] = "analyze";
    Phase["Implement"] = "implement";
    Phase["Verify"] = "verify";
    Phase["Release"] = "release";
})(Phase || (Phase = {}));
/** Ordered array of all phases for iteration */
export const PHASE_ORDER = [
    Phase.Init,
    Phase.Discover,
    Phase.Specify,
    Phase.Clarify,
    Phase.Design,
    Phase.Tasks,
    Phase.Analyze,
    Phase.Implement,
    Phase.Verify,
    Phase.Release,
];
/** Maps each phase to the required file(s) that must exist before advancing past it */
export const PHASE_REQUIRED_FILES = {
    [Phase.Init]: ["CONSTITUTION.md"],
    [Phase.Discover]: [],
    [Phase.Specify]: ["SPECIFICATION.md"],
    [Phase.Clarify]: ["SPECIFICATION.md"],
    [Phase.Design]: ["DESIGN.md"],
    [Phase.Tasks]: ["TASKS.md"],
    [Phase.Analyze]: ["ANALYSIS.md"],
    [Phase.Implement]: ["CHECKLIST.md"],
    [Phase.Verify]: ["VERIFICATION.md"],
    [Phase.Release]: [],
};
/** All MCP tool names — prefixed with sdd_ */
export const TOOL_NAMES = {
    // Pipeline tools
    INIT: "sdd_init",
    DISCOVER: "sdd_discover",
    WRITE_SPEC: "sdd_write_spec",
    CLARIFY: "sdd_clarify",
    WRITE_DESIGN: "sdd_write_design",
    WRITE_TASKS: "sdd_write_tasks",
    RUN_ANALYSIS: "sdd_run_analysis",
    ADVANCE_PHASE: "sdd_advance_phase",
    // Utility tools
    GET_STATUS: "sdd_get_status",
    GET_TEMPLATE: "sdd_get_template",
    WRITE_BUGFIX: "sdd_write_bugfix",
    CHECK_SYNC: "sdd_check_sync",
    SCAN_CODEBASE: "sdd_scan_codebase",
    AMEND: "sdd_amend",
    // Transcript automation tools
    IMPORT_TRANSCRIPT: "sdd_import_transcript",
    AUTO_PIPELINE: "sdd_auto_pipeline",
    BATCH_TRANSCRIPTS: "sdd_batch_transcripts",
    // Input & Conversion
    IMPORT_DOCUMENT: "sdd_import_document",
    BATCH_IMPORT: "sdd_batch_import",
    FIGMA_TO_SPEC: "sdd_figma_to_spec",
    // Quality & Validation
    CHECKLIST: "sdd_checklist",
    VERIFY_TASKS: "sdd_verify_tasks",
    COMPLIANCE_CHECK: "sdd_compliance_check",
    CROSS_ANALYZE: "sdd_cross_analyze",
    VALIDATE_EARS: "sdd_validate_ears",
    // Diagrams & Visualization
    GENERATE_DIAGRAM: "sdd_generate_diagram",
    GENERATE_ALL_DIAGRAMS: "sdd_generate_all_diagrams",
    GENERATE_USER_STORIES: "sdd_generate_user_stories",
    FIGMA_DIAGRAM: "sdd_figma_diagram",
    // Infrastructure as Code
    GENERATE_IAC: "sdd_generate_iac",
    VALIDATE_IAC: "sdd_validate_iac",
    GENERATE_DOCKERFILE: "sdd_generate_dockerfile",
    // Dev Environment
    SETUP_LOCAL_ENV: "sdd_setup_local_env",
    SETUP_CODESPACES: "sdd_setup_codespaces",
    GENERATE_DEVCONTAINER: "sdd_generate_devcontainer",
    // Integration & Export
    CREATE_BRANCH: "sdd_create_branch",
    EXPORT_WORK_ITEMS: "sdd_export_work_items",
    CREATE_PR: "sdd_create_pr",
    IMPLEMENT: "sdd_implement",
    RESEARCH: "sdd_research",
    // Documentation
    GENERATE_DOCS: "sdd_generate_docs",
    GENERATE_API_DOCS: "sdd_generate_api_docs",
    GENERATE_RUNBOOK: "sdd_generate_runbook",
    GENERATE_ONBOARDING: "sdd_generate_onboarding",
    // Metrics
    METRICS: "sdd_metrics",
    // Testing
    GENERATE_TESTS: "sdd_generate_tests",
    VERIFY_TESTS: "sdd_verify_tests",
    // Property-Based Testing
    GENERATE_PBT: "sdd_generate_pbt",
    // Turnkey
    TURNKEY_SPEC: "sdd_turnkey_spec",
    // Checkpointing
    CHECKPOINT: "sdd_checkpoint",
    RESTORE: "sdd_restore",
    LIST_CHECKPOINTS: "sdd_list_checkpoints",
    // All Documentation
    GENERATE_ALL_DOCS: "sdd_generate_all_docs",
    // Ecosystem
    CHECK_ECOSYSTEM: "sdd_check_ecosystem",
    // Model Routing
    MODEL_ROUTING: "sdd_model_routing",
    // Context Tiering
    CONTEXT_STATUS: "sdd_context_status",
    // RBAC
    CHECK_ACCESS: "sdd_check_access",
};
// Input & Conversion
export const TOOL_IMPORT_DOCUMENT = "sdd_import_document";
export const TOOL_BATCH_IMPORT = "sdd_batch_import";
export const TOOL_FIGMA_TO_SPEC = "sdd_figma_to_spec";
// Quality & Validation
export const TOOL_CHECKLIST = "sdd_checklist";
export const TOOL_VERIFY_TASKS = "sdd_verify_tasks";
export const TOOL_COMPLIANCE_CHECK = "sdd_compliance_check";
export const TOOL_CROSS_ANALYZE = "sdd_cross_analyze";
export const TOOL_VALIDATE_EARS = "sdd_validate_ears";
// Diagrams & Visualization
export const TOOL_GENERATE_DIAGRAM = "sdd_generate_diagram";
export const TOOL_GENERATE_ALL_DIAGRAMS = "sdd_generate_all_diagrams";
export const TOOL_GENERATE_USER_STORIES = "sdd_generate_user_stories";
export const TOOL_FIGMA_DIAGRAM = "sdd_figma_diagram";
// Infrastructure as Code
export const TOOL_GENERATE_IAC = "sdd_generate_iac";
export const TOOL_VALIDATE_IAC = "sdd_validate_iac";
export const TOOL_GENERATE_DOCKERFILE = "sdd_generate_dockerfile";
// Dev Environment
export const TOOL_SETUP_LOCAL_ENV = "sdd_setup_local_env";
export const TOOL_SETUP_CODESPACES = "sdd_setup_codespaces";
export const TOOL_GENERATE_DEVCONTAINER = "sdd_generate_devcontainer";
// Integration & Export
export const TOOL_CREATE_BRANCH = "sdd_create_branch";
export const TOOL_EXPORT_WORK_ITEMS = "sdd_export_work_items";
export const TOOL_CREATE_PR = "sdd_create_pr";
export const TOOL_IMPLEMENT = "sdd_implement";
export const TOOL_RESEARCH = "sdd_research";
// Documentation
export const TOOL_GENERATE_DOCS = "sdd_generate_docs";
export const TOOL_GENERATE_API_DOCS = "sdd_generate_api_docs";
export const TOOL_GENERATE_RUNBOOK = "sdd_generate_runbook";
export const TOOL_GENERATE_ONBOARDING = "sdd_generate_onboarding";
// Metrics
export const TOOL_METRICS = "sdd_metrics";
/** Template names matching files in templates/ directory */
export const TEMPLATE_NAMES = [
    "constitution",
    "specification",
    "design",
    "tasks",
    "analysis",
    "bugfix",
    "sync_report",
    "research",
    "data_model",
    "checklist",
    "cross_analysis",
    "work_items",
    "verification",
    "compliance",
    "user_stories",
    "api_docs",
    "runbook",
    "onboarding",
    "terraform",
    "dockerfile",
    "devcontainer",
    "journey",
];
/** EARS pattern names */
export const EARS_PATTERNS = [
    "ubiquitous",
    "event_driven",
    "state_driven",
    "optional",
    "unwanted",
    "complex",
    "unknown",
];
/** Package manifest files for tech stack detection, in priority order */
export const PACKAGE_MANIFESTS = [
    "package.json",
    "requirements.txt",
    "pyproject.toml",
    "go.mod",
    "Cargo.toml",
    "pom.xml",
    "build.gradle",
];
