// specky-sdd/dist/cli/lib/settings-merger.js
/**
 * settings-merger.ts — Deep-merge Claude Code hooks into .claude/settings.json.
 *
 * Claude Code reads `.claude/settings.json` for hooks, permissions, etc.
 * We must merge (not overwrite) since users may have their own hooks/permissions.
 */
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
/** Tools that Specky agents need pre-authorized (avoids per-invocation prompts). */
export const SPECKY_REQUIRED_ALLOWS = [
    // Native Claude tools used by agents
    "Read",
    "Glob",
    "Grep",
    "Edit",
    "Write",
    "MultiEdit",
    "WebFetch",
    "WebSearch",
    "Task",
    // Shell — Specky agents and hooks need these utilities
    "Bash(bash:*)",
    "Bash(sh:*)",
    "Bash(git:*)",
    "Bash(npm:*)",
    "Bash(npx:*)",
    "Bash(node:*)",
    "Bash(ls:*)",
    "Bash(mkdir:*)",
    "Bash(rm:*)",
    "Bash(cp:*)",
    "Bash(mv:*)",
    "Bash(touch:*)",
    "Bash(chmod:*)",
    "Bash(cat:*)",
    "Bash(head:*)",
    "Bash(tail:*)",
    "Bash(wc:*)",
    "Bash(find:*)",
    "Bash(grep:*)",
    "Bash(sed:*)",
    "Bash(awk:*)",
    "Bash(jq:*)",
    "Bash(bc:*)",
    "Bash(pip:*)",
    "Bash(pip3:*)",
    "Bash(python:*)",
    "Bash(python3:*)",
    // All Specky MCP tools
    "mcp__specky__*",
];
function readSettings(path) {
    if (!existsSync(path))
        return {};
    try {
        const raw = readFileSync(path, "utf8");
        if (!raw.trim())
            return {};
        return JSON.parse(raw);
    }
    catch (err) {
        throw new Error(`[specky] Failed to parse ${path}: ${err.message}`);
    }
}
/**
 * Merge strategy: for each event (PreToolUse/PostToolUse/Stop),
 *   - find existing groups by matcher; if matcher already exists with identical
 *     command set, skip. Otherwise append.
 *   - never delete user-authored groups.
 */
function mergeHooks(existing, incoming) {
    const merged = { ...existing };
    for (const [event, groups] of Object.entries(incoming)) {
        const existingGroups = merged[event] ?? [];
        const out = [...existingGroups];
        for (const g of groups) {
            const match = out.find((e) => e.matcher === g.matcher);
            if (!match) {
                out.push(g);
                continue;
            }
            // Append any hook commands not already present
            for (const h of g.hooks) {
                const exists = match.hooks.some((x) => x.type === h.type && x.command === h.command);
                if (!exists)
                    match.hooks.push(h);
            }
        }
        merged[event] = out;
    }
    return merged;
}
/**
 * Merge permissions.allow — union with existing, dedupe.
 * Never removes user-authored entries.
 */
function mergePermissions(existing, requiredAllows) {
    const base = existing ?? {};
    const currentAllow = Array.isArray(base.allow) ? base.allow : [];
    const set = new Set(currentAllow);
    let added = 0;
    for (const a of requiredAllows) {
        if (!set.has(a)) {
            set.add(a);
            added++;
        }
    }
    return {
        merged: { ...base, allow: [...set] },
        added,
    };
}
export function mergeClaudeHooks(targets, claudeHooks, opts) {
    const path = targets.claude.settingsJson;
    const existing = readSettings(path);
    const mergedHooks = mergeHooks(existing.hooks ?? {}, claudeHooks);
    const addedEvents = Object.keys(mergedHooks).filter((k) => !existing.hooks || !existing.hooks[k]);
    const { merged: mergedPerms, added: addedPermissions } = mergePermissions(existing.permissions, SPECKY_REQUIRED_ALLOWS);
    const final = {
        ...existing,
        hooks: mergedHooks,
        permissions: mergedPerms,
    };
    if (!opts.dryRun) {
        mkdirSync(dirname(path), { recursive: true });
        writeFileSync(path, JSON.stringify(final, null, 2) + "\n", "utf8");
    }
    return { written: !opts.dryRun, path, addedEvents, addedPermissions };
}
export function loadClaudeHooksManifest(manifestPath) {
    if (!existsSync(manifestPath)) {
        throw new Error(`[specky] claude-hooks.json not found at ${manifestPath}. ` +
            `Did the build step run? (npm run build)`);
    }
    const raw = readFileSync(manifestPath, "utf8");
    return JSON.parse(raw);
}
export { resolve };
