// specky-sdd/dist/services/template-engine.js
/**
 * TemplateEngine — Template loading, variable replacement, YAML frontmatter.
 */
import { join } from "node:path";
import { readFile } from "node:fs/promises";
import { fileURLToPath } from "node:url";
import { dirname } from "node:path";
import { TEMPLATE_NAMES } from "../constants.js";
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const TEMPLATES_DIR = join(__dirname, "..", "templates");
export class TemplateEngine {
    /** Optional path to project-local templates directory (from .specky/config.yml) */
    customTemplatesDir;
    constructor(_fileManager, customTemplatesPath) {
        this.customTemplatesDir = customTemplatesPath
            ? join(_fileManager.workspaceRoot, customTemplatesPath)
            : "";
    }
    /**
     * Load and render a template with variable replacement.
     */
    async render(templateName, context) {
        const template = await this.loadTemplate(templateName);
        return this.replaceVariables(template, context);
    }
    /**
     * Render a template with YAML frontmatter prepended.
     */
    async renderWithFrontmatter(templateName, context) {
        const frontmatter = this.generateFrontmatter(context);
        const body = await this.render(templateName, context);
        return `${frontmatter}\n${body}`;
    }
    /**
     * Get raw template content without any processing.
     */
    async getTemplate(templateName) {
        return this.loadTemplate(templateName);
    }
    /**
     * Replace {{variable}} placeholders with context values.
     * Unknown variables become [TODO: variable].
     * Supports {{#each items}}...{{/each}} for arrays.
     */
    replaceVariables(template, context) {
        let result = template;
        // Handle {{#each key}}...{{/each}} blocks
        const eachRegex = /\{\{#each\s+(\w+)\}\}([\s\S]*?)\{\{\/each\}\}/g;
        result = result.replace(eachRegex, (_match, key, body) => {
            const value = context[key];
            if (Array.isArray(value)) {
                return value.map((item) => body.replace(/\{\{this\}\}/g, item)).join("");
            }
            return `[TODO: ${key}]`;
        });
        // Handle {{variable}} replacements
        result = result.replace(/\{\{(\w+)\}\}/g, (_match, key) => {
            const value = context[key];
            if (value === undefined) {
                return `[TODO: ${key}]`;
            }
            if (Array.isArray(value)) {
                return value.join(", ");
            }
            return value;
        });
        return result;
    }
    /**
     * Generate YAML frontmatter block.
     */
    generateFrontmatter(context) {
        const lines = ["---"];
        const fields = [
            ["title", context.title],
            ["feature_id", context.feature_id],
            ["version", context.version || "1.0.0"],
            ["date", context.date || new Date().toISOString().split("T")[0]],
            ["author", context.author || "SDD Pipeline"],
            ["status", context.status || "Draft"],
        ];
        for (const [key, value] of fields) {
            if (value !== undefined) {
                if (Array.isArray(value)) {
                    lines.push(`${key}: [${value.map((v) => `"${v}"`).join(", ")}]`);
                }
                else {
                    lines.push(`${key}: "${value}"`);
                }
            }
        }
        lines.push("---");
        return lines.join("\n");
    }
    /**
     * Load raw template — checks .specky/templates/ first, then built-in templates/.
     * This implements T-066: customizable templates via project-local directory.
     */
    async loadTemplate(name) {
        if (!TEMPLATE_NAMES.includes(name)) {
            throw new Error(`Unknown template: "${name}". Available: ${TEMPLATE_NAMES.join(", ")}`);
        }
        // Map template name to file: sync_report -> sync-report.md
        const fileName = name.replace(/_/g, "-") + ".md";
        // Try custom templates directory first (T-066)
        if (this.customTemplatesDir) {
            try {
                return await readFile(join(this.customTemplatesDir, fileName), "utf-8");
            }
            catch {
                // Custom template not found — fall through to built-in
            }
        }
        // Fall back to built-in templates
        try {
            return await readFile(join(TEMPLATES_DIR, fileName), "utf-8");
        }
        catch {
            throw new Error(`Template file not found: ${fileName}. Ensure templates/ directory exists.`);
        }
    }
}
