// specky-sdd/dist/services/test-result-parser.js
/**
 * TestResultParser — Parses test runner output into a normalized TestResult[].
 * Supports: Vitest JSON, pytest JSON, JUnit XML.
 */
export class TestResultParser {
    /**
     * Auto-detect format and parse.
     */
    parse(raw) {
        const trimmed = raw.trim();
        if (trimmed.startsWith("<"))
            return this.parseJUnit(trimmed);
        try {
            const json = JSON.parse(trimmed);
            return this.parseJson(json);
        }
        catch {
            return [];
        }
    }
    /** Vitest JSON: { testResults: [{ name, status, message? }] } or flat array */
    parseJson(json) {
        if (!json || typeof json !== "object")
            return [];
        // Vitest format: { testResults: [...] }
        if ("testResults" in json && Array.isArray(json["testResults"])) {
            return json["testResults"].flatMap((suite) => {
                if (!suite || typeof suite !== "object")
                    return [];
                const s = suite;
                const tests = s["assertionResults"] ?? s["tests"] ?? [];
                return tests.map((t) => this.normalizeJsonTest(t));
            });
        }
        // pytest JSON: { tests: [...] }
        if ("tests" in json && Array.isArray(json["tests"])) {
            return json["tests"].map((t) => this.normalizeJsonTest(t));
        }
        // Flat array
        if (Array.isArray(json)) {
            return json.map((t) => this.normalizeJsonTest(t));
        }
        return [];
    }
    normalizeJsonTest(t) {
        const name = String(t["name"] ?? t["nodeid"] ?? t["fullName"] ?? "unknown");
        const statusStr = String(t["status"] ?? t["outcome"] ?? "").toLowerCase();
        const passed = statusStr === "passed" || statusStr === "pass";
        const failureMessages = t["failureMessages"];
        const error = failureMessages
            ? failureMessages.join(" ").slice(0, 500)
            : t["message"] ? String(t["message"]).slice(0, 500) : undefined;
        const duration_ms = typeof t["duration"] === "number" ? t["duration"] : undefined;
        return {
            test_name: name,
            passed,
            ...(error ? { error } : {}),
            ...(duration_ms !== undefined ? { duration_ms } : {}),
        };
    }
    /** JUnit XML: <testcase name="..." time="..."><failure>...</failure></testcase> */
    parseJUnit(xml) {
        const results = [];
        // Self-closing alternative must come first to avoid greedy [^>]* consuming the trailing slash
        const testcaseRegex = /<testcase\s([^>]*)\/>|<testcase\s([^>]*)(?<!\/)>([\s\S]*?)<\/testcase>/g;
        let m;
        while ((m = testcaseRegex.exec(xml)) !== null) {
            const attrs = m[1] ?? m[2] ?? ""; // m[1]: self-closing attrs; m[2]: open-tag attrs
            const inner = m[3] ?? ""; // m[3]: content between open and close tags
            const name = this.xmlAttr(attrs, "name") ?? this.xmlAttr(attrs, "classname") ?? "unknown";
            const timeStr = this.xmlAttr(attrs, "time");
            const duration_ms = timeStr ? Math.round(parseFloat(timeStr) * 1000) : undefined;
            const failureMatch = inner.match(/<failure[^>]*>([\s\S]*?)<\/failure>/);
            const errorMatch = inner.match(/<error[^>]*>([\s\S]*?)<\/error>/);
            const errorText = ((failureMatch?.[1] ?? errorMatch?.[1] ?? "").slice(0, 500).trim()) || undefined;
            results.push({
                test_name: name,
                passed: !failureMatch && !errorMatch,
                ...(errorText ? { error: errorText } : {}),
                ...(duration_ms !== undefined ? { duration_ms } : {}),
            });
        }
        return results;
    }
    xmlAttr(attrs, key) {
        const m = attrs.match(new RegExp(`${key}="([^"]*)"`, "i"));
        return m?.[1];
    }
}
