// specky-sdd/dist/services/work-item-exporter.js
export class WorkItemExporter {
    fileManager;
    constructor(fileManager) {
        this.fileManager = fileManager;
    }
    async export(platform, _specDir, featureDir, _includeSubtasks, _options) {
        const tasksContent = await this.fileManager.readSpecFile(featureDir, "TASKS.md");
        const specContent = await this.fileManager.readSpecFile(featureDir, "SPECIFICATION.md");
        const tasks = this.parseTasks(tasksContent, specContent);
        const featureNumber = featureDir.match(/(\d{3})/)?.[1] || "000";
        const featureName = featureDir.replace(/.*\d{3}-/, "");
        const items = tasks.map(t => ({
            task_id: t.id,
            title: t.title,
            description: t.description,
            traces_to: t.traces_to,
            effort: t.effort,
            dependencies: t.dependencies,
            acceptance_criteria: t.acceptance_criteria,
        }));
        const routing = this.getRoutingInstructions(platform);
        return {
            platform,
            items,
            metadata: { feature_number: featureNumber, feature_name: featureName, total_items: items.length, generated_at: new Date().toISOString() },
            routing_instructions: routing,
        };
    }
    toGitHubPayloads(items, featureNumber) {
        return items.map(item => ({
            title: `[${item.task_id}] ${item.title}`,
            body: `## Description\n${item.description}\n\n## Acceptance Criteria\n${item.acceptance_criteria.map(ac => `- [ ] ${ac}`).join("\n")}\n\n## Traces To\n${item.traces_to.join(", ")}\n\n## Effort\n${item.effort || "Not estimated"}`,
            labels: ["sdd", `feature/${featureNumber}`, item.effort ? `effort:${item.effort}` : ""].filter(Boolean),
            assignees: [],
        }));
    }
    toAzureBoardsPayloads(items, areaPath, iterationPath) {
        return items.map(item => ({
            title: `[${item.task_id}] ${item.title}`,
            description: `<h2>Description</h2><p>${item.description}</p><h2>Traces To</h2><p>${item.traces_to.join(", ")}</p>`,
            work_item_type: "Task",
            area_path: areaPath,
            iteration_path: iterationPath,
            tags: ["sdd", item.effort || ""].filter(Boolean),
            acceptance_criteria: item.acceptance_criteria.map(ac => `<li>${ac}</li>`).join(""),
        }));
    }
    toJiraPayloads(items, projectKey) {
        return items.map(item => ({
            summary: `[${item.task_id}] ${item.title}`,
            description: `h2. Description\n${item.description}\n\nh2. Acceptance Criteria\n${item.acceptance_criteria.map(ac => `* ${ac}`).join("\n")}\n\nh2. Traces To\n${item.traces_to.join(", ")}`,
            issue_type: "Task",
            project_key: projectKey,
            labels: ["sdd", item.effort || ""].filter(Boolean),
            priority: "Medium",
        }));
    }
    parseTasks(tasksContent, specContent) {
        const tasks = [];
        const taskRegex = /^-\s+\[[ x]\]\s+(T\d{3})\s+(?:\[P\]\s+)?(?:\[(US\d+)\]\s+)?(.+)/gm;
        let match;
        while ((match = taskRegex.exec(tasksContent)) !== null) {
            const id = match[1];
            const story = match[2] || "";
            const title = match[3].trim();
            // Extract REQ IDs referenced in the task or linked story
            const reqIds = [];
            const reqRegex = /REQ-[A-Z]+-\d{3}/g;
            let reqMatch;
            while ((reqMatch = reqRegex.exec(title)) !== null)
                reqIds.push(reqMatch[0]);
            // If no direct REQ ref, look in spec for the story
            if (reqIds.length === 0 && story) {
                const storySection = specContent.match(new RegExp(`${story}[\\s\\S]*?(?=##|$)`));
                if (storySection) {
                    while ((reqMatch = reqRegex.exec(storySection[0])) !== null)
                        reqIds.push(reqMatch[0]);
                }
            }
            tasks.push({ id, title, description: title, traces_to: reqIds, dependencies: [], acceptance_criteria: [] });
        }
        return tasks;
    }
    getRoutingInstructions(platform) {
        switch (platform) {
            case "github": return { mcp_server: "github", tool_name: "create_issue", note: "Call GitHub MCP create_issue for each item" };
            case "azure_boards": return { mcp_server: "azure-devops", tool_name: "create_work_item", note: "Call Azure DevOps MCP create_work_item for each item" };
            case "jira": return { mcp_server: "jira", tool_name: "create_issue", note: "Call Jira MCP create_issue for each item" };
        }
    }
}
