// specky-sdd/dist/tools/response-builder.js
/**
 * ResponseBuilder — Enriches tool responses with phase context, educational content,
 * handoff information, and parallel execution hints.
 */
import { MethodologyGuide } from "../services/methodology.js";
import { DependencyGraph } from "../services/dependency-graph.js";
import { routingEngine } from "../utils/routing-helper.js";
import { buildDefaultContextSummary } from "../utils/context-helper.js";
/**
 * Build an enriched tool response with phase context and educational content.
 */
export function buildToolResponse(toolName, result, currentPhase, phases, options) {
    // Phase context
    const progress = MethodologyGuide.getProgressIndicator(currentPhase, phases);
    const phaseContext = {
        current_phase: currentPhase,
        phase_progress: progress.progress_bar,
        phases_completed: progress.completed_phases,
        completion_percent: progress.percent_complete,
    };
    // Educational content
    const toolExplanation = MethodologyGuide.getToolExplanation(toolName);
    const phaseExplanation = MethodologyGuide.getPhaseExplanation(currentPhase);
    // Handoff context (if phase transition occurred)
    let handoff;
    if (options?.completedPhase && options?.nextPhase) {
        const nextExplanation = MethodologyGuide.getPhaseExplanation(options.nextPhase);
        handoff = {
            completed_phase: options.completedPhase,
            next_phase: options.nextPhase,
            artifacts_produced: options.artifactsProduced || [],
            summary_of_work: options.summaryOfWork || `Completed ${options.completedPhase} phase.`,
            what_comes_next: nextExplanation.what,
            methodology_note: nextExplanation.sdd_principle,
        };
    }
    // Parallel execution hints
    const deps = DependencyGraph.getDependencies(toolName);
    const executionPlan = DependencyGraph.getExecutionPlan(currentPhase);
    const parallelOpportunities = {
        can_run_now: deps.parallel_with,
        must_wait_for: deps.requires.filter(r => !progress.completed_phases.includes(r)),
        explanation: executionPlan.next_steps.length > 0
            ? `Next: ${executionPlan.next_steps[0].description}${executionPlan.next_steps[0].parallel ? " (can run in parallel)" : ""}`
            : "Pipeline complete.",
    };
    return {
        ...result,
        phase_context: phaseContext,
        handoff,
        parallel_opportunities: parallelOpportunities,
        educational_note: toolExplanation.why_it_matters,
        methodology_tip: phaseExplanation.sdd_principle,
        model_routing_hint: routingEngine.getHint(currentPhase),
        context_load_summary: buildDefaultContextSummary(),
    };
}
/**
 * Convenience wrapper: loads state from StateMachine and enriches a tool response.
 * This is the primary function tools should call — handles state loading internally.
 */
export async function enrichResponse(toolName, result, stateMachine, specDir, options) {
    try {
        const state = await stateMachine.loadState(specDir);
        return buildToolResponse(toolName, result, state.current_phase, state.phases, options);
    }
    catch {
        // If state can't be loaded, return result as-is with minimal enrichment
        return {
            ...result,
            educational_note: MethodologyGuide.getToolExplanation(toolName).why_it_matters,
        };
    }
}
/**
 * Enrich a stateless tool response (tools without spec_dir or stateMachine).
 * Provides educational content without phase context.
 */
export function enrichStateless(toolName, result) {
    const toolExplanation = MethodologyGuide.getToolExplanation(toolName);
    return {
        ...result,
        educational_note: toolExplanation.why_it_matters,
        common_mistakes: toolExplanation.common_mistakes,
        sdd_context: "This is a utility tool that operates independently of the SDD pipeline phase. It can be called at any time to support your workflow.",
    };
}
/**
 * Build a validation error response when a tool is called in the wrong phase.
 */
export function buildPhaseError(toolName, currentPhase, expectedPhases, errorMessage) {
    const phaseExplanation = MethodologyGuide.getPhaseExplanation(currentPhase);
    return {
        error: "phase_validation_failed",
        tool: toolName,
        current_phase: currentPhase,
        expected_phases: expectedPhases,
        message: errorMessage,
        fix: `Complete the ${currentPhase} phase first. ${phaseExplanation.how}`,
        methodology_note: phaseExplanation.sdd_principle,
    };
}
