// specky-sdd/dist/cli/commands/init.js
/**
 * init.ts — `specky init` — install Specky assets into the current workspace.
 */
import { copyFileSync, existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { resolve } from "node:path";
import { VERSION } from "../../constants.js";
import { copyToClaude, copyToCopilot, summarizeCopy, writeInstallLock, } from "../lib/asset-copier.js";
import { detectIde } from "../lib/ide-detect.js";
import { packageRoot, sourcePaths, targetPaths } from "../lib/paths.js";
import { loadClaudeHooksManifest, mergeClaudeHooks } from "../lib/settings-merger.js";
import { writeMcpRegistration } from "../lib/mcp-writer.js";
import { writeVscodeSettings } from "../lib/vscode-settings-writer.js";
import { writeGitignoreBlock } from "../lib/gitignore-writer.js";
function installClaude(ctx) {
    console.log("[specky init] Installing to .claude/ …");
    const r = copyToClaude(ctx.pkg, ctx.targets, ctx.copyOpts);
    console.log(summarizeCopy(".claude", r));
    try {
        const hooks = loadClaudeHooksManifest(ctx.src.claudeHooksManifest);
        const merge = mergeClaudeHooks(ctx.targets, hooks, { dryRun: ctx.dryRun });
        console.log(`  settings.json: ${merge.written ? "merged" : "dry-run"} at ${merge.path}`);
        if (merge.addedPermissions > 0) {
            console.log(`  permissions:   ${merge.addedPermissions} allow rules added (specky MCP + native tools)`);
        }
    }
    catch (err) {
        console.warn(`  ⚠️  ${err.message}`);
    }
    const mcp = writeMcpRegistration(ctx.targets.shared.claudeMcp, {
        dryRun: ctx.dryRun,
        serverName: "specky",
        useVscodeSchema: false,
    });
    console.log(`  .mcp.json: ${mcp.action}`);
    console.log("");
    return r;
}
function installCopilot(ctx) {
    console.log("[specky init] Installing to .github/ …");
    const r = copyToCopilot(ctx.pkg, ctx.targets, ctx.copyOpts);
    console.log(summarizeCopy(".github", r));
    const mcp = writeMcpRegistration(ctx.targets.shared.vscodeMcp, {
        dryRun: ctx.dryRun,
        serverName: "specky",
        useVscodeSchema: true,
    });
    console.log(`  .vscode/mcp.json: ${mcp.action}`);
    const vscodeSettingsPath = resolve(ctx.workspace, ".vscode/settings.json");
    const vs = writeVscodeSettings(vscodeSettingsPath, { dryRun: ctx.dryRun });
    if (vs.addedKeys.length > 0) {
        console.log(`  .vscode/settings.json: +${vs.addedKeys.length} key(s) (${vs.addedKeys.join(", ")})`);
    }
    else {
        console.log(`  .vscode/settings.json: already configured`);
    }
    console.log("");
    return r;
}
function writeSpeckyMeta(ctx, resolvedIde) {
    if (ctx.dryRun)
        return;
    mkdirSync(ctx.targets.shared.specky, { recursive: true });
    const configDest = resolve(ctx.targets.shared.specky, "config.yml");
    if (!existsSync(configDest) && existsSync(ctx.src.configYml)) {
        copyFileSync(ctx.src.configYml, configDest);
        console.log(`[specky init] Wrote ${configDest}`);
    }
    const meta = {
        version: VERSION,
        ide: resolvedIde,
        installed_at: new Date().toISOString(),
    };
    writeFileSync(resolve(ctx.targets.shared.specky, "install.json"), JSON.stringify(meta, null, 2) + "\n", "utf8");
}
function formatGitignoreAction(action) {
    if (action === "created")
        return "created";
    if (action === "updated")
        return "updated";
    return "already present";
}
function ensureGitignore(ctx) {
    // Add/refresh the Specky-managed block in .gitignore so vendored assets
    // (agents, prompts, skills, hooks regenerated by `specky install`) aren't
    // committed. Idempotent — safe to call on every install/upgrade.
    const r = writeGitignoreBlock(ctx.workspace, { dryRun: ctx.dryRun });
    const verb = formatGitignoreAction(r.action);
    const suffix = r.linesAdded > 0 ? `+${r.linesAdded} lines` : "no changes";
    console.log(`[specky init] .gitignore: ${verb} (${suffix})`);
}
function printHeader(opts, workspace, resolvedIde, detected) {
    const autoNote = opts.ide === "auto" || !opts.ide ? " (auto-detected)" : "";
    console.log(`[specky init] Version ${VERSION}`);
    console.log(`[specky init] Workspace: ${workspace}`);
    console.log(`[specky init] IDE target: ${resolvedIde}${autoNote}`);
    if (detected.signals.length) {
        console.log(`[specky init] Detected signals: ${detected.signals.join(", ")}`);
    }
    if (opts.dryRun)
        console.log(`[specky init] DRY RUN — no files will be written`);
    console.log("");
}
function printFooter() {
    console.log("");
    console.log("[specky init] ✅ Done. Next steps:");
    console.log("  • In Claude Code or VS Code: restart the session to activate MCP + hooks");
    console.log("  • Start the pipeline: invoke @specky-onboarding (Copilot) or /specky-onboarding (Claude)");
    console.log("  • Run `npx specky doctor` anytime to validate install integrity");
}
export async function runInit(opts) {
    const workspace = opts.workspace ?? process.cwd();
    const pkg = packageRoot();
    const ctx = {
        workspace,
        pkg,
        targets: targetPaths(workspace),
        src: sourcePaths(pkg),
        copyOpts: { force: opts.force, dryRun: opts.dryRun },
        dryRun: opts.dryRun,
    };
    const detected = detectIde(workspace);
    const resolvedIde = !opts.ide || opts.ide === "auto" ? detected.recommendation : opts.ide;
    printHeader(opts, workspace, resolvedIde, detected);
    const results = [];
    if (resolvedIde === "claude" || resolvedIde === "both") {
        results.push(installClaude(ctx));
    }
    if (resolvedIde === "copilot" || resolvedIde === "both") {
        results.push(installCopilot(ctx));
    }
    // rc.14: When target is copilot-only, strip hooks from .claude/settings.json
    // to prevent Copilot from cross-reading Claude Code lifecycle hooks.
    // Copilot reads .claude/settings.json and treats SessionStart/UserPromptSubmit
    // as PreToolUse, causing spurious "Blocked by Pre-Tool Use hook" errors.
    if (resolvedIde === "copilot" && !ctx.dryRun) {
        const claudeSettings = ctx.targets.claude.settingsJson;
        if (existsSync(claudeSettings)) {
            try {
                const raw = readFileSync(claudeSettings, "utf8");
                const parsed = JSON.parse(raw);
                if (parsed.hooks) {
                    delete parsed.hooks;
                    writeFileSync(claudeSettings, JSON.stringify(parsed, null, 2) + "\n", "utf8");
                    console.log("  .claude/settings.json: stripped hooks (Copilot cross-read fix)");
                }
            }
            catch {
                // ignore — user may have non-JSON content
            }
        }
    }
    writeSpeckyMeta(ctx, resolvedIde);
    ensureGitignore(ctx);
    const lockPath = writeInstallLock(ctx.targets, results, VERSION, ctx.copyOpts);
    console.log(`[specky init] Integrity manifest: ${lockPath}`);
    printFooter();
    return 0;
}
